// @_koii/task-manager/dist/esm/extensions/orca/index.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { OrcaPulse } from "@orcacompute/orca-node-client";
import { TASK_ID } from "@_koii/namespace-wrapper";
import { getPodSpec } from "./podSpec.js";
import { getTaskManager } from "../../main/taskManager.js";
class OrcaClient {
    constructor(client = null, isInitializing = false, isInitialized = false, rootCA = null) {
        this.client = client;
        this.isInitializing = isInitializing;
        this.isInitialized = isInitialized;
        this.rootCA = rootCA;
    }
    initialize(imageURL, podSpec, rootCA) {
        return __awaiter(this, void 0, void 0, function* () {
            console.log("Initializing Orca client");
            const taskManager = yield getTaskManager();
            taskManager.registerExtension("orca", {
                percent: 0.01,
                publicKey: "ORCA",
            });
            if (this.isInitializing || this.isInitialized) {
                return;
            }
            this.isInitializing = true;
            this.rootCA = rootCA;
            try {
                this.client = new OrcaPulse();
                yield this.client.initialize(imageURL, TASK_ID, "ws://127.0.0.1:3003", podSpec, "healthz", rootCA);
                this.isInitialized = true;
                console.log("Orca client initialized");
            }
            catch (error) {
                let errorMessage;
                if (error instanceof Error) {
                    errorMessage = error.message;
                }
                else {
                    errorMessage = JSON.stringify(error);
                }
                this.client = null;
                throw new Error(`Orca client initialization failed: ${errorMessage}`);
            }
            finally {
                this.isInitializing = false;
            }
        });
    }
}
const orcaClient = new OrcaClient();
export function initializeOrcaClient(config) {
    return __awaiter(this, void 0, void 0, function* () {
        const { imageURL, customPodSpec, rootCA } = config;
        console.log({ customPodSpec });
        orcaClient.initialize(imageURL, yield getPodSpec(customPodSpec, rootCA), rootCA);
    });
}
export function getOrcaClient() {
    return __awaiter(this, void 0, void 0, function* () {
        return new Promise((resolve, reject) => {
            if (orcaClient.isInitialized) {
                resolve(orcaClient.client);
                return;
            }
            const checkInterval = setInterval(() => {
                if (orcaClient.isInitialized) {
                    clearInterval(checkInterval);
                    resolve(orcaClient.client);
                }
            }, 500);
            setTimeout(() => {
                clearInterval(checkInterval);
                reject(new Error("OrcaClient initialization timed out."));
            }, 60000);
        });
    });
}
